### ISSUE TITLE
Optimize version conflict strategy

### ISSUE THREAD

**Issue by otakustay** [yarnpkg/yarn#1150]
**Do you want to request a _feature_ or report a _bug_?**

feature

**What is the current behavior?**

Suppose we have dependencies below:
- `a` depends on `underscore@1.7.x`
- `b` depends on `underscore@1.7.x`
- `c` depends on `underscore@1.5.x`
- `main` depends on a, b & c

after `yarn install` in main module, the node_modules tree could be:

``` text
node_modules
├── a
│   └── node_modules
│       └── underscore <-- 1.7.x
├── b
│   └── node_modules
│       └── underscore <-- 1.7.x
├── c
└── underscore <-- 1.5.x
```

**What is the expected behavior?**

It's not an optimized result since `underscore@1.7.x` is used more times than `1.5.x`, it could be better to place `1.7.x` under the top node_modules directory and only put a sub `1.5.x` in `c/node_modules`

In frontend environment we really care about file sizes, but `--flat` flag may introduce compatibility issues, so to optimize the version conflict strategy can provide considerable benifits

**Please mention your node.js, yarn and operating system version.**
- Node: v6.8.0
- yarn: v0.15.1
- OS: macOS 10.12

**Comment by destroyerofbuilds** [comment]
@otakustay and @Daniel15, I believe this issue will be mitigated, if not entirely resolved by #2676

**Comment by gsklee** [comment]
@destroyerofbuilds indeed, thanks!
